### PR TITLE
chore(ci): add `cargo deny` ci and config

### PR DESCRIPTION
**Pull request by antbern**
# Pull Request Template

## Motivation and Context

This runs `cargo-deny` in the CI. It checks the licenses of all dependencies to make sure they do not validate our use. Also checks against the RUSTSEC security advisory database (same as `cargo audit`) and 

## Summary

Provide a short summary of changes in this pull request.

- Add cargo deny config file with a list of accepted licenses (reviewers please double-check if reasonable)
- Add CI job that runs cargo deny.
- Mark all crates except `sdk` and `api_types` as non-publishable since they are only used internally to generate bindings and artifacts.

## Checklist

Please ensure that your PR meets the following requirements:

### General

- [x] Code follows project style and guidelines.
- [x] No redundant or duplicate code.
- [x] No added new dependencies without clear justification and approval.

### Documentation

- [ ] Added/updated relevant documentation (e.g., README, inline comments).
- [ ] Added a CHANGELOG entry for major changes.

### Testing

- [ ] Added relevant test cases, leveraging tools like `rstest` or similar for reusable tests.
- [ ] Tests pass locally.

### Build & CI/CD

- [x] Confirmed no CI/CD pipeline issues.
- [x] Updated build scripts where necessary.

## Comments to Reviewer

Please provide any comments or points of consideration for the reviewers, such as specific areas of focus, potential edge cases, or context that would aid the review process.

### Reviewer Responsibilities

- [ ] Code adheres to style and guidelines.
- [ ] All tests are appropriately covered and they pass.
- [ ] Consider performance and/or security impacts.
- [ ] Constructive feedback is provided.

### CODE REVIEW

**Review by hauju — APPROVED** [review]

File: deny.toml [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    "Apache-2.0",
+    "Apache-2.0 WITH LLVM-exception",
+    "Unicode-3.0",	
+	"Unicode-DFS-2016",
+	"ISC",
+	"MPL-2.0",
+	"CC0-1.0",
+    "BSD-3-Clause",
+    "BSD-2-Clause",
+	"GPL-3.0",
```
Comment (kristifidani) [inline deny.toml:103]: I did a quick search on these and I found these 2 might need a second look or not.

* GPL-3.0
> Using GPL v3 code may introduce safety risks because users must be enabled to deploy modified versions on their devices. Legal compliance: Producers must ensure that they comply with all the license requirements when they distribute their software, either as part of a device or separately. If you allow GPL-3.0, any derivative work must also be GPL-3.0. Some projects avoid GPL to prevent licensing conflicts.

* OpenSSL 
> ring previously depended on OpenSSL but now avoids it in favor of BoringSSL, Rustls, or custom implementations.

Well I have not much info on these but we can also leave it as it is for now. We will have time to review these later 💯

Comment (antbern) [reply deny.toml:103]: Yeah I'm aware that these two might be problematic. We need to look at it separately, but at least now we know about them 💯 And the only two deps that are covered by this is `ring` (used in `rustls` among others) and `kdbx-rs` (only one with GPL-3.0)
